### PR TITLE
[expo-av][iOS] fix race conditions

### PR DESCRIPTION
**Pull request by hirbod**
# Why

Essentially, the thought process is that the Video unmounts before the React Native bridge is able to send out the unloadAsync() command in some circumstances, which leads to:

```log
NSInternalInconsistencyException: An instance 0x28225af10 of class AVPlayerLayer was deallocated while key value observers were still registered with it. Current observation info: <NSKeyValueObservationInfo 0x2832b0d00> (
<NSKeyValueObservance 0x282264540: Observer: 0x10783e7a0, Key path: readyForDisplay, Options: <New: YES, Old: NO, Prior: NO> Context: 0x0, Property: 0x283a19560>
)
```

Thanks once again to @Pickleboyonline 

# How

Instead of calling `pauseImmediately` in the dealloc method, we're calling the Obj-C equivalent for `unloadAsync()` and call `setSource:nil`

Although we are not quite sure if this will make the crash reports disappear, it is worth a try, as the de-allocation should be cleaner this way

# Test Plan

Tested on couple of real devices and inside the simulator, no regressions.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by bbarthec — APPROVED** [review]
Looks fine to me 👍
Thank you for this fix 💪

**Comment by barthap** [comment]
This code is crashing for me, when unmounting a `<Video>` component in ImagePicker NCL screen (video preview disappears): https://github.com/expo/expo/blob/3bc9c491a49471bbd7115ba977e6ef8a5b76cf59/apps/native-component-list/src/screens/ImagePicker/ImageOrVideoPreview.tsx#L13-L19

Screenshot 2022-07-05 at 11 11 38
Screenshot 2022-07-05 at 11 11 42

**Comment by barthap** [comment]
But I think I've just found a fix, trying it out now 😉

**Comment by hirbod** [comment]
Alright, let's revert this PR for now

**Comment by hirbod** [comment]
@Pickleboyonline guess another path has to be taken

**Comment by Pickleboyonline** [comment]
**TL;DR:** 
> Use `useLayoutEffect` instead of `useEffect` when executing the `unloadAsync()` Expo Video function on cleanup.
> This will ensure that setSource:nil has started before React Native has queued an unmount.
> We could also update the `componentWillUnmount` method in the Video class component to remove the unloading
> burden from `expo-av` users and remove confusion.

Ok @hirbod I think I found a way to do this without making any native changes. With `delloc`, you are apparently not supposed to but any thread dependent code in it (see here: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/mmPractical.html#//apple_ref/doc/uid/TP40004447-SW13 )

The docs recommend that you perform cleanup on scarce resources **before** you release the object. React Native makes this difficult because it does not expose any UIViewController that we can take advantage of (UIViewController contains lifecycle methods we could have used). After some research, I wasn't able to come up with a solution that either didn't introduce major changes to the Obj C code.

To avoid that I searched for a JS fix. I remembered that `componentWillUnmount` from the React Class era could help here and after some Stack Overflow I saw that `useLayoutEffect` can have the same result as the `willMount`/`willUnmount`. Calling `unloadAsync()` in the cleanup did indeed work with no error.

Therefore, I believe the best way to resolve this issue is to update the `componentWillUnmount()` within the Video class component to remove this burden from expo-av users.
